config/kernel-blkdev.m4

```diff
@@ -294,6 +294,27 @@ AC_DEFUN([ZFS_AC_KERNEL_BLKDEV_BDEV_WHOLE], [
 	])
 ])
 
+dnl #
+dnl # 5.13 API change
+dnl # blkdev_get_by_path() no longer handles ERESTARTSYS
+dnl #
+dnl # Unfortunately we're forced to rely solely on the kernel version
+dnl # number in order to determine the expected behavior.  This was an
+dnl # internal change to blkdev_get_by_dev(), see commit a8ed1a0607.
+dnl #
+AC_DEFUN([ZFS_AC_KERNEL_BLKDEV_GET_ERESTARTSYS], [
+	AC_MSG_CHECKING([whether blkdev_get_by_path() handles ERESTARTSYS])
+	AS_VERSION_COMPARE([$LINUX_VERSION], [5.13.0], [
+		AC_MSG_RESULT(yes)
+		AC_DEFINE(HAVE_BLKDEV_GET_ERESTARTSYS, 1,
+			[blkdev_get_by_path() handles ERESTARTSYS])
+	],[
+		AC_MSG_RESULT(no)
+	],[
+		AC_MSG_RESULT(no)
+	])
+])
+
```

Comment:
I know of at least one case where people have been packaging newer (>=5.13) kernels that incorporate a revert patch for a8ed1a0607 (the 5.13 commit which removed the `ERESTARTSYS` retry path in the block layer).

In such a situation, this simple kernel-version-based config check would conclude that the block layer `ERESTARTSYS` retry path does not exist in the block layer for those kernels (i.e. `HAVE_BLKDEV_GET_ERESTARTSYS` is not defined).

Would this circumstance (having both the <5.13 block layer retry logic *plus* the zfs >=5.13 internal retry logic conditional on `#ifndef HAVE_BLKDEV_GET_ERESTARTSYS`) cause any problems? Is this something that warrants testing?

Reply:
That's good to know.  Functionally I don't think that should cause a problem, we'd just potentially do some additional retries.  Performance wise there would now be a small 10ms delay in that retry case where with the patched kernel and a version of ZFS without this patch it'd retry immediately.  It'd be ideal if we could reliably detect these kernels but since it was an internal function change I didn't see an easy way to detect it.

If you have one of these patched kernels handy it'd be worth running the test suite to make sure.

Reply:
> It'd be ideal if we could reliably detect these kernels but since it was an internal function change I didn't see an easy way to detect it.

Yeah... I was able to think of a couple *really heinous* ways to detect it. 😄 Like:
- Inspecting the source code used to build the target kernel. (But the source code isn't always readily available. And you'd need to ensure it actually does correspond to the target kernel headers etc. And it's just generally awful.)
- Inspecting the compiled binary of the target kernel. (Which is just horrendous in so many ways. Ultimately you'd have to disassemble and figure out the control flow in `blkdev_get_by_dev`. Which is not necessarily hard to do by hand, but would be horrible to have to automate for every possible compiler version and optimization permutation. Oh and also this is, yet again, just generally awful.)

Both of those ideas are so hideous that it's hard to imagine they'd *ever* be a good idea, even in a case where the change in question *really actually mattered* (like if it was potentially a big deal for data integrity or something). 😬

The only other way I can think of to do the detection, that wouldn't be totally horrible, happens to be eminently impractical. So basically this idea would be to have a small test kernel module that implements a block device driver, compiled at config time and loaded into the kernel to do a runtime check rather than a static one. (The main impracticality obviously arises from the running host kernel and build target kernel not being guaranteed to be the same—and in fact *normally* not being the same.) The test module would mostly just consist of an `open` handler that:
- returns `-ERESTARTSYS` the first time it's called for a given device node
- returns `0` on subsequent calls for that same device node

And then the config script, having compiled and loaded that module, would do a pair of `open(2)` syscalls on a test block device and see whether the result comes back (a) as a pair of successes; or (b) if the first call gives `errno=512`, indicating that the retry path isn't existent. And of course this would be impractical because it'd necessitate:
- the build host kernel being the same as the build target kernel
- the config script literally loading a kernel module temporarily (and having the required permissions to do that)
- having a libc that doesn't sanitize-out non-sane `errno` values like those in the >=512 space (and while it may be true that current upstream glibc does let them through, I have literally no idea what other libc implementations do)

Maybe the last point doesn't matter, since you could hypothetically do the syscall manually rather than using the libc `open(2)` wrapper function... But really, it's not like *any* of this matters since it's all stupid and impractical. 😀

---

> If you have one of these patched kernels handy it'd be worth running the test suite to make sure.

Okay yeah, I'll see if I can do that. I'll see if I can spin up a VM, do a quick kernel + zfs build (with both the kernel 1-commit revert and this PR integrated), and then run the test suite. Come to think of it, I should probably already have set up a VM for this sort of thing in the past...

Reply:
Right, all of those options really aren't going to be workable.  Let's go forward with this admittedly crude check.

Reply:
Just to make sure I went ahead and tested this with a pre-5.13 kernel which includes the `blkdev_get_by_dev()` retry logic and then undefined `HAVE_BLKDEV_GET_ERESTARTSYS`.  As expected everything worked and it wasn't a problem.  @jgottula if your testing turns up something unexpected for this odd case just let me know.  In the meanwhile let's go ahead and get this merged.